service/history/workflow/transaction_impl.go

```diff
@@ -251,18 +251,31 @@ func (t *TransactionImpl) SetWorkflowExecution(
 func PersistWorkflowEvents(
 	ctx context.Context,
 	shard shard.Context,
-	workflowEvents *persistence.WorkflowEvents,
+	workflowEventsSlice ...*persistence.WorkflowEvents,
 ) (int64, error) {
 
-	if len(workflowEvents.Events) == 0 {
-		return 0, nil // allow update workflow without events
-	}
+	var totalSize int64
+	for _, workflowEvents := range workflowEventsSlice {
+		if len(workflowEvents.Events) == 0 {
+			continue // allow update workflow without events
+		}
 
-	firstEventID := workflowEvents.Events[0].EventId
-	if firstEventID == common.FirstEventID {
-		return persistFirstWorkflowEvents(ctx, shard, workflowEvents)
+		firstEventID := workflowEvents.Events[0].EventId
+		if firstEventID == common.FirstEventID {
+			size, err := persistFirstWorkflowEvents(ctx, shard, workflowEvents)
+			if err != nil {
+				return 0, err
+			}
+			totalSize += size
+		} else {
+			size, err := persistNonFirstWorkflowEvents(ctx, shard, workflowEvents)
+			if err != nil {
+				return 0, err
+			}
+			totalSize += size
```

Comment:
nit: you could move 
```
			if err != nil {
				return 0, err
			}
			totalSize += size
``` out of the if..else block and reduce duplicated code.

Reply:
my main concern is not the duplicated code, but the `:=` from go, sometimes the var can be shadow-ed, causing unwanted behavior, e.g. https://github.com/temporalio/temporal/pull/1187

err defined here: https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL109

err shadow-ed here: https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL112

shadow-ed err assignment here:https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL127

err check here: https://github.com/temporalio/temporal/pull/1187/files#diff-72a05e1636a2bcea965072880e2c16672336b99edab8cf54c0add86e5877c09eL135

basically err is defined outside of the if block, got shadowed & assigned within the if block; later when the error check is done from outside the if block, the err is *not* the updated err from within the if block.